src/ch341a.c

```diff
@@ -0,0 +1,467 @@
+
+  if (!CH341USBTransfer(pgm, LIBUSB_ENDPOINT_OUT, pkt, size + 1)) {
+    fprintf(stderr, "Error: failed to transfer data to CH341\n");
+    return -1;
+  }
+
+  if (!CH341USBTransfer(pgm, LIBUSB_ENDPOINT_IN, pkt, size)) {
+    fprintf(stderr, "Error: failed to transfer data from CH341\n");
+    return -1;
+  }
+
+  for (i = 0; i < size; i++) out[i] = swap_byte(pkt[i]);
+
+  return size;
+}
+
+static int ch341a_spi_program_enable(const PROGRAMMER *pgm, const AVRPART *p)
+{
+  unsigned char res[4];
+  unsigned char cmd[4];
+  memset(cmd, 0, sizeof(cmd));
+  memset(res, 0, sizeof(res));
+
+  cmd[0] = 0;
+  pmsg_trace("ch341a_program_enable() %s\n", p->op[AVR_OP_PGM_ENABLE]);
+
+  if (p->op[AVR_OP_PGM_ENABLE] == NULL) {
+    pmsg_error("program enable instruction not defined for part \"%s\"\n",
+              p->desc);
+    return -1;
+  }
+  avr_set_bits(p->op[AVR_OP_PGM_ENABLE], cmd);
+  pgm->cmd(pgm, cmd, res);
+
+  pmsg_debug("program_enable(): sending command. Resp = %x %x %x %x \n",
+            (int)res[0], (int)res[1], (int)res[2], (int)res[3]);
+  // check for sync character
+  if (res[2] != cmd[1]) return -2;
+  return 0;
+}
+
+static int ch341a_spi_chip_erase(const struct programmer_t *pgm,
+                                 const AVRPART *p)
+{
+  unsigned char cmd[4];
+  unsigned char res[4];
+  if (p->op[AVR_OP_CHIP_ERASE] == NULL) {
+    pmsg_error("chip erase instruction not defined for part \"%s\"\n", p->desc);
+    return -1;
+  }
+  memset(cmd, 0, sizeof(cmd));
+  avr_set_bits(p->op[AVR_OP_CHIP_ERASE], cmd);
+  pgm->cmd(pgm, cmd, res);
+  usleep(p->chip_erase_delay);
+  pgm->initialize(pgm, p);
+  return 0;
+}
+
+
+void ch341a_initpgm(PROGRAMMER * pgm)
+{
+  strcpy(pgm->type, "ch341a");
+  /*
+  * mandatory functions
+  */
+  pgm->initialize     = ch341a_initialize;
+  pgm->display        = ch341a_display;
+  pgm->enable         = ch341a_enable;
+  pgm->disable        = ch341a_disable;
+  pgm->program_enable = ch341a_spi_program_enable;
+  pgm->chip_erase     = ch341a_spi_chip_erase;
+  pgm->cmd            = ch341a_spi_cmd;
+  pgm->spi            = ch341a_spi;
+  pgm->open           = ch341a_open;
+  pgm->close          = ch341a_close;
+  pgm->read_byte      = avr_read_byte_default;
+  pgm->write_byte     = avr_write_byte_default;
+
+  /*
+  * optional functions
+  */
+
+  //pgm->paged_write    = ch341a_spi_paged_write;
+  //pgm->paged_load     = ch341a_spi_paged_load;
+  pgm->setup          = ch341a_setup;
+  pgm->teardown       = ch341a_teardown;
+}
+
+/* Interface - management */
+static void ch341a_setup(PROGRAMMER *pgm)
+{
+  if ((pgm->cookie = malloc(sizeof(struct pdata))) == 0) {
+    pmsg_error("ch341a_setup(): Out of memory allocating private data\n");
+    exit(1);
+  }
+  memset(pgm->cookie, 0, sizeof(struct pdata));
+}
+
+static void ch341a_teardown(PROGRAMMER *pgm)
+{
+  free(pgm->cookie);
+}
+
+/* Dummy functions */
+static void ch341a_disable(const PROGRAMMER *pgm)
+{
+  /* Do nothing. */
```

Comment:
Also vvv minor. There is a comment above that already says these are dummy functions. It is very clear they don't do anything except resolving the symbol for the linker, so so need for the comment `/* Do nothing */`